lib/async.js

```diff
@@ -265,19 +274,18 @@ module.exports = function resolve(x, options, callback) {
         if (dirs.length === 0) return cb(null, undefined);
         var dir = dirs[0];
 
-        isDirectory(dir, isdir);
+        isDirectory(path.dirname(dir), isdir);
```

Comment:
when is the dirname not going to be a directory? the intention here, i believe, is to see if `dir` points to a file or not. It seems like hoisting the `path.join` prevents that from being determined.

Reply:
The `dirname` isn't a directory when it doesn't exist. From what I gather from #190 and #116, this code was only meant to speed up the resolution by checking whether the n_m folder *existed* before doing the file stat calls.

Checking whether the parent of the subpath (ie `/n_m/foo` for `foo/hello`) exists has the same behaviours: if `foo/hello` can be resolved, then `/n_m/foo` will necessarily be a directory.

Reply:
In other words:

| request | partial resolution | isDirectory (before) | isDirectory (after) |
| --- | --- | --- | --- |
| `foo` | `/n_m/foo` | `/n_m` | `/n_m` |
| `foo/bar` | `/n_m/foo/bar` | `/n_m` | `/n_m/foo` |
| `@foo/bar` | `/n_m/@foo/bar` | `/n_m` | `/n_m/@foo` |

The effect will be the same before and after: the two stat calls that the code would have made before #190 (taking the first line as an example, one to check whether `/n_m/foo.js` exists and another to check whether `/n_m/foo/index.js` exists) won't be necessary because `resolve` is able to find out that the directory supposed to contain them doesn't even exist.

Reply:
I see. Do you think that this perf improvement could also be split out into a separate PR (or even just a separate commit prior to your addition)?

Reply:
In this PR it's not a perf improvement, it's necessary to make my PR work with the changes brought by #190 (because with my PR we don't have access to `n_m` paths anymore in `loadNodeModules` - instead we only have `/n_m/<subpath>`). I guess I could revert it and re-apply it, but I'm not sure it would be much clearer 🤔

Reply:
I'm mainly concerned about an unintentional breakage, and my hope would be to release this particular change as a patch prior to releasing this PR as a minor.

Reply:
I think #190 has only been released in the 2.0.0-next.0 branch, so for the purpose of the patch release the `path.dirname` thing can be removed entirely. Do you want me to open a separate PR against the `1.x` branch?

Reply:
#192 was the PR into the 1.x branch.